core/chains/evm/txm/storage/inmemory_store.go

```diff
@@ -0,0 +1,333 @@
+	tx.AttemptCount++
+	m.UnconfirmedTransactions[txNonce].Attempts = append(m.UnconfirmedTransactions[txNonce].Attempts, attempt.DeepCopy())
+
+	return nil
+}
+
+func (m *InMemoryStore) CountUnstartedTransactions() int {
+	m.RLock()
+	defer m.RUnlock()
+
+	return len(m.UnstartedTransactions)
+}
+
+func (m *InMemoryStore) CreateEmptyUnconfirmedTransaction(nonce uint64, gasLimit uint64) (*types.Transaction, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	m.txIDCount++
+	emptyTx := &types.Transaction{
+		ID:                m.txIDCount,
+		ChainID:           m.chainID,
+		Nonce:             &nonce,
+		FromAddress:       m.address,
+		ToAddress:         common.Address{},
+		Value:             big.NewInt(0),
+		SpecifiedGasLimit: gasLimit,
+		CreatedAt:         time.Now(),
+		State:             types.TxUnconfirmed,
+	}
+
+	if _, exists := m.UnconfirmedTransactions[nonce]; exists {
+		return nil, fmt.Errorf("an unconfirmed tx with the same nonce already exists: %v", m.UnconfirmedTransactions[nonce])
+	}
+
+	m.UnconfirmedTransactions[nonce] = emptyTx
+	m.Transactions[emptyTx.ID] = emptyTx
+
+	return emptyTx.DeepCopy(), nil
+}
+
+func (m *InMemoryStore) CreateTransaction(txRequest *types.TxRequest) *types.Transaction {
+	m.Lock()
+	defer m.Unlock()
+
+	m.txIDCount++
+
+	tx := &types.Transaction{
+		ID:                m.txIDCount,
+		IdempotencyKey:    txRequest.IdempotencyKey,
+		ChainID:           m.chainID,
+		FromAddress:       m.address,
+		ToAddress:         txRequest.ToAddress,
+		Value:             txRequest.Value,
+		Data:              txRequest.Data,
+		SpecifiedGasLimit: txRequest.SpecifiedGasLimit,
+		CreatedAt:         time.Now(),
+		State:             types.TxUnstarted,
+		Meta:              txRequest.Meta,
+		MinConfirmations:  txRequest.MinConfirmations,
+		PipelineTaskRunID: txRequest.PipelineTaskRunID,
+		SignalCallback:    txRequest.SignalCallback,
+	}
+
+	if len(m.UnstartedTransactions) == maxQueuedTransactions {
+		m.lggr.Warnf("Unstarted transactions queue for address: %v reached max limit of: %d. Dropping oldest transaction: %v.",
+			m.address, maxQueuedTransactions, m.UnstartedTransactions[0])
+		delete(m.Transactions, m.UnstartedTransactions[0].ID)
+		m.UnstartedTransactions = m.UnstartedTransactions[1:maxQueuedTransactions]
+	}
+
+	txCopy := tx.DeepCopy()
+	m.Transactions[txCopy.ID] = txCopy
+	m.UnstartedTransactions = append(m.UnstartedTransactions, txCopy)
+	return tx
+}
+
+func (m *InMemoryStore) FetchUnconfirmedTransactionAtNonceWithCount(latestNonce uint64) (txCopy *types.Transaction, unconfirmedCount int) {
+	m.RLock()
+	defer m.RUnlock()
+
+	tx := m.UnconfirmedTransactions[latestNonce]
+	if tx != nil {
+		txCopy = tx.DeepCopy()
+	}
+	unconfirmedCount = len(m.UnconfirmedTransactions)
+	return
+}
+
+func (m *InMemoryStore) MarkTransactionsConfirmed(latestNonce uint64) ([]*types.Transaction, []uint64, error) {
```

Comment:
Think this method name might be misleading since it also handles marking re-org'd transactions as unconfirmed

Reply:
How about `MarkConfirmedAndReorgedTransactions` ?

Reply:
I'm ok with that